tests/repo_finder/test_repo_finder_deps_dev.py

```diff
@@ -0,0 +1,192 @@
+
+"""This module tests the deps.dev repo finder."""
+
+import pytest
+from packageurl import PackageURL
+from pytest_httpserver import HTTPServer
+
+from macaron.repo_finder.repo_finder_deps_dev import DepsDevRepoFinder
+from macaron.repo_finder.repo_finder_enums import RepoFinderInfo
+
+
+def test_find_repo_url_failure(deps_dev_service_mock: dict) -> None:
+    """Test find repo function."""
+    purl = PackageURL.from_string(f"pkg:pypi/example{deps_dev_service_mock['api']}")
+    result, outcome = DepsDevRepoFinder().find_repo(purl)
+    assert not result
+    assert outcome == RepoFinderInfo.DDEV_API_ERROR
+
+
+@pytest.mark.parametrize(
+    ("data", "expected_outcome"),
+    [
+        ('{"foo": "bar"}', RepoFinderInfo.DDEV_JSON_INVALID),
+        ('{"links": [{"url": 1}]}', RepoFinderInfo.DDEV_NO_URLS),
+        ('{"links": [{"url": "test://test.test"}]}', RepoFinderInfo.DDEV_NO_VALID_URLS),
+    ],
+)
+def test_find_repo_links_failures(
+    httpserver: HTTPServer, deps_dev_service_mock: dict, data: str, expected_outcome: RepoFinderInfo
+) -> None:
+    """Test invalid links."""
+    purl = PackageURL.from_string("pkg:pypi/example@2")
+    target_url = (
+        f"/{deps_dev_service_mock['api']}/{deps_dev_service_mock['purl']}/pkg:{purl.type}/{purl.name}@{purl.version}"
+    )
+
+    httpserver.expect_request(target_url).respond_with_data(data)
+    result, outcome = DepsDevRepoFinder().find_repo(purl)
+
+    assert not result
+    assert outcome == expected_outcome
+
+
+def test_find_repo_success(httpserver: HTTPServer, deps_dev_service_mock: dict) -> None:
+    """Test repo finder success."""
+    purl = PackageURL.from_string("pkg:pypi/example@2")
+    target_url = (
+        f"/{deps_dev_service_mock['api']}/{deps_dev_service_mock['purl']}/pkg:{purl.type}/{purl.name}@{purl.version}"
+    )
+
+    httpserver.expect_request(target_url).respond_with_data('{"links": [{"url": "http://github.com/oracle/macaron"}]}')
+    result, outcome = DepsDevRepoFinder().find_repo(purl)
+
+    assert result
+    assert outcome == RepoFinderInfo.FOUND
+
+
+@pytest.mark.parametrize(
+    ("repo_url", "server_url", "data"),
+    [
+        ("http::::://130/test", "", ""),
+        ("http://github.com/oracle/macaron", "", ""),
+        ("", "/oracle/macaron", "INVALID JSON"),
+    ],
+)
+def test_get_project_info_failures(
+    httpserver: HTTPServer, deps_dev_service_mock: dict, repo_url: str, server_url: str, data: str
+) -> None:
+    """Test get project info failures."""
+    if not repo_url:
+        repo_url = f"{deps_dev_service_mock['base_scheme']}://{deps_dev_service_mock['base_netloc']}{server_url}"
+
+    if server_url:
+        target_url = f"/{deps_dev_service_mock['api']}/projects/{deps_dev_service_mock['base_hostname']}{server_url}"
+        httpserver.expect_request(target_url).respond_with_data(data)
+
+    assert not DepsDevRepoFinder().get_project_info(repo_url)
```

Comment:
I think there are 2 cases of failures we are testing in this function
- The Repository URL is given, and not valid
- The repository URL is available (from pytest-httpserver) but the result JSON is invalid

Could we have 2 separated test functions for each case, so that we can simplify the if conditions on the test inputs.

Reply:
I have separated the two test cases.